prediction_market_agent/agents/microchain_agent/app.py

```diff
@@ -48,20 +48,38 @@
 ALLOW_STOP = False
 
 
-def run_agent(agent: Agent, iterations: int, model: str) -> None:
-    with openai_costs(model) as costs:
+def run_agent(agent: Agent, iterations: int) -> None:
+    maybe_initialize_long_term_memory()
+    with openai_costs(st.session_state.model) as costs:
         with st.spinner("Agent is running..."):
             for _ in range(iterations):
                 agent.run(iterations=1, resume=True)
         st.session_state.running_cost += costs.cost
 
 
-def execute_reasoning(agent: Agent, reasoning: str, model: str) -> None:
-    with openai_costs(model) as costs:
-        agent.execute_command(f'Reasoning("{reasoning}")')
-        display_new_history_callback(agent)  # Run manually after `execute_command`
+def on_click_run_agent() -> None:
+    maybe_initialize_agent(system_prompt, bootstrap)
+    run_agent(
+        agent=st.session_state.agent,
+        iterations=int(iterations),
+    )
+    save_last_turn_history_to_memory(st.session_state.agent)
+
+
+def on_click_user_reasoning() -> None:
+    maybe_initialize_agent(system_prompt, bootstrap)
+
+    with openai_costs(st.session_state.model) as costs:
+        st.session_state.agent.execute_command(f'Reasoning("{user_reasoning}")')
         st.session_state.running_cost += costs.cost
 
+    save_last_turn_history_to_memory(st.session_state.agent)
```

Comment:
Above, this is call after running the agent, here it is called before running the agent. What's the correct way?

Reply:
@evangriffiths could you please help me understand the logic here? https://github.com/gnosis/prediction-market-agent/blob/db46c34c87aed6e6b1f7582a7771b363c2fb7a7f/prediction_market_agent/agents/microchain_agent/app.py#L268-L281
My questions:
-> Why do you call `agent.run` after `agent.execute_command`?  From the [Agent file](https://github.com/galatolofederico/microchain/blob/20320cded714f12a11e7f237671faec4174b55b1/microchain/engine/agent.py#L4), I understand that `agent.execute_command` and `agent.run` behave similarly, hence only one is needed.
-> If you agree with the above, I would suggest removing the subsequent call to `run_agent` after the `execute_reasoning` is called (line 275 approx).

Reply:
That was meant to be a UX improvement. Before the user did:
1. type in some reasoning, hit enter -> the agent executes the reasoning, adds it to its history, and this is displayed in the app
2. the user clicks on 'run agent'

Since I assumed (1) is always followed by (2), I just merged them for convenience.

Reply:
Nice, thanks. So I changed the button to `add reasoning and run the agent` for clarity.